WalletWasabi/WabiSabi/Backend/Banning/CoinVerifier.cs

```diff
@@ -21,87 +23,96 @@ public CoinVerifier(CoinJoinIdStore coinJoinIdStore, CoinVerifierApiClient apiCl
+	public async Task<IEnumerable<CoinVerifyResult>> VerifyCoinsAsync(IEnumerable<Coin> coinsToCheck, CancellationToken cancellationToken)
 	{
-		var before = DateTimeOffset.UtcNow;
-
 		using CancellationTokenSource cancellationTokenSource = new(TimeSpan.FromSeconds(30));
 		using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
 
-		var lastChangeId = Whitelist.ChangeId;
-		Whitelist.RemoveAllExpired(WabiSabiConfig.ReleaseFromWhitelistAfter);
-
-		var scriptsToCheck = new HashSet<Script>();
-		var innocentsCounter = 0;
+		// Booting up the results with the default value - ban: no, remove: yes.
+		Dictionary<Coin, CoinVerifyResult> coinVerifyItems = coinsToCheck.ToDictionary(coin => coin, coin => new CoinVerifyResult(coin, ShouldBan: false, ShouldRemove: true));
 
+		// Building up the task list.
+		List<Task<CoinVerifyResult>> tasks = new();
 		foreach (var coin in coinsToCheck)
 		{
-			if (CheckIfAlreadyVerified(coin))
-			{
-				innocentsCounter++;
-				yield return new CoinVerifyInfo(false, coin);
-			}
-			else
+			if (!CoinVerifyItems.TryGetValue(coin, out var item))
 			{
-				scriptsToCheck.Add(coin.ScriptPubKey);
+				// If the coin was not scheduled try to quickly schedule it - it should not happen.
+				Logger.LogWarning($"Trying to re-schedule coin '{coin.Outpoint}' for verification.");
+
+				// Quickly re-scheduling the missing items.
+				ScheduleVerification(coin, cancellationToken, TimeSpan.Zero);
+				if (!CoinVerifyItems.TryGetValue(coin, out item))
+				{
+					// This should not happen.
+					Logger.LogError($"Coin '{coin.Outpoint}' cannot be re-scheduled for verification. The coin will be removed from the round.");
+					continue;
+				}
 			}
+
+			tasks.Add(item.TaskCompletionSource.Task);
 		}
 
-		Logger.LogInfo($"{innocentsCounter} out of {coinsToCheck.Count()} utxo is already verified in Round({roundId}).");
-		await foreach (var response in CoinVerifierApiClient.VerifyScriptsAsync(scriptsToCheck, linkedCts.Token))
+		try
 		{
-			bool shouldBanUtxo = CheckForFlags(response.ApiResponseItem);
-
-			// Find all coins with the same script (address reuse).
-			foreach (var coin in coinsToCheck.Where(c => c.ScriptPubKey == response.ScriptPubKey))
+			do
 			{
-				if (!shouldBanUtxo)
-				{
-					Whitelist.Add(coin.Outpoint);
-				}
+				var completedTask = await Task.WhenAny(tasks).WaitAsync(linkedCts.Token).ConfigureAwait(false);
+				tasks.Remove(completedTask);
+				var result = await completedTask.WaitAsync(linkedCts.Token).ConfigureAwait(false);
 
-				yield return new CoinVerifyInfo(shouldBanUtxo, coin);
+				// The verification task fulfilled its purpose - clean up.
+				CoinVerifyItems.TryRemove(result.Coin, out var _);
+
+				// Update the default value with the real result.
+				coinVerifyItems[result.Coin] = result;
 			}
+			while (tasks.Any());
```

Comment:
Related to this comment https://github.com/molnard/WalletWasabi/pull/47#issuecomment-1386774794

If we use normal `while` cycle (instead of do-while), the mentioned exception doesn't come.

Reply:
Fixed - good catch!